### ISSUE TITLE
Range generator has non standard behaviour 

### ISSUE THREAD

**Issue by NicolasDP** [vincenthz/smoke#4]
Hi, we have been using your great crate in order to write our property test (great job btw, it's real smooth).

While using it we found that writing range generator didn't have the expected behaviour (though, at least, the one we would expect from the rust's standard library). Indeed the following range `1..2` yield either `1` or `2`. While in the rust standard library we would expect to always have `1` since the upper bound is excluded, see: https://doc.rust-lang.org/std/ops/struct.Range.html

How would you feel if we were to help you fix this so we can use [`RangeBoudns`](https://doc.rust-lang.org/std/ops/trait.RangeBounds.html) instead of doing a custom Range ?

I think the problem can be addressed two ways:

# Call `RangeBounds::end_bound`

Here, instead of using `self.0.start`:
 https://github.com/vincenthz/smoke/blob/e41e2e5b0ba333cf362376cf33d956ce61eb2130/smoke/src/generator/numerical.rs#L41

we would use `self.0.end_bound()` to get the upper bound:

# Replace the `NumRange` generator by a generic Range

i.e. instead of

https://github.com/vincenthz/smoke/blob/e41e2e5b0ba333cf362376cf33d956ce61eb2130/smoke/src/generator/numerical.rs#L30

we could try to see how to use:

```rust
pub struct Range<T>(Box<dyn RangeBound<T>>);
```

**Comment by vincenthz** [comment]
I think that sounds reasonable for the inclusive/exclusive bound change;

not totally sure that RangeBounds works in this case, but if it does I think that make sense as a general range mechanism for smoke

**Comment by NicolasDP** [comment]
you are right RangeBounds will not work. We will need to go with something more specific `{ start: Bounds<T>, end: Bounds<T> }`
